calculator_test.go

```diff
@@ -1,24 +1,133 @@
+func TestDivide(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name        string
+		a           float64
+		b           []float64
+		want        float64
+		errExpected bool
+	}{
+		{"Dividing a number by 1 yields the same number", 10, []float64{1}, 10, false},
+		{"Dividing a number by 1 multiple times yields the same number", 99, []float64{1, 1, 1}, 99, false},
+		{"Dividing a number by zero returns an error", 312, []float64{0}, 0, true},
+		{"Dividing a number by multiple numbers with zero being one of them returns an error", 50, []float64{1, 5, 0, 200}, 0, true},
+		{"Dividing a zero by zero returns an error", 0, []float64{0}, 0, true},
+		{"Dividing a number by multiple larger numbers yields a fraction", 1, []float64{2, 2, 2}, 0.125, false},
+		{"Dividing a number by a large number and a fraction", 100, []float64{0.1, 10}, 100, false},
+		{"Dividing a number by a much larger number yields a small fraction", 100, []float64{10000000}, 0.00001, false},
+	}
+	for _, c := range testCases {
+		got, err := Divide(c.a, c.b...)
+		errReceived := err != nil
+		if c.errExpected != errReceived {
+			t.Fatalf("%s: unexpected error status %v", c.name, err)
+		}
+		if !errReceived && c.want != got {
+			t.Errorf("%s: want %f, got %f", c.name, c.want, got)
+		}
+	}
+}
+
+func TestSqrt(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name        string
+		a, want     float64
+		errExpected bool
+	}{
+		{"The square root of 1 is 1", 1, 1, false},
+		{"The square root of zero is zero", 0, 0, false},
+		{"The square root of a two digit number", 25, 5, false},
+		{"The square root of a large number", 1947028, 1395.3594518976106, false},
+		{"The square root of a small fraction", 0.00000002, 0.0001414213562373095, false},
+		{"Attempting to calculate the square root of a negative number returns an error", -25, 0, true},
+	}
+	for _, c := range testCases {
+		got, err := Sqrt(c.a)
+		errReceived := err != nil
+		if c.errExpected != errReceived {
+			t.Fatalf("%s: unexpected error status %v", c.name, err)
+		}
+		if !errReceived && c.want != got {
+			t.Errorf("%s: want %f, got %f", c.name, c.want, got)
+		}
+	}
+}
+func TestEvaluate(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name        string
+		expression  string
+		want        float64
+		errExpected bool
+	}{
+		{"A simple valid addition expression", "1 + 1", 2, false},
+		{"A simple valid multiplication expression", "10 * 10.0", 100, false},
+		{`Trying to evaluate an unimplemented operation returns an error`, "2 % 5", 0, true},
+		{"Trying to evaluate an invalid operation returns an error", "51x5.3", 0, true},
+		{"An expression with both operands containing fractions", "43.75 / 3.5", 12.5, false},
+		{"An expression containing no spaces", "2/2", 0, true},
+		{"An expression with whitepsace around the second operand", "1000/    10   \n	  ", 0, true},
+		{`Trying to evaluate an expression with more than two operands returns an error`, "8 * 3 / 9", 0, true},
```

Comment:
With `fmt.Sscanf`, there don't seem to be a way to return an error if expression extends beyond the first operation and the two operands. Or am I missing something?

Reply:
That's a good point. Suppose you used `fmt.Scanf` instead, which takes a reader, and then you look at the reader to see if there's anything left in it which `Scanf` didn't consume. That would be an error.

Reply:
I assumed you meant `fmt.Fscanf` since `Scanf` reads from `io.Stdin` only (by wrapping `Fscanf`).

Reply:
Right!

Reply:
Done!